// web3-eth-contract/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCreateAccessListParams = exports.isWeb3ContractContext = exports.isContractInitOptions = exports.getEstimateGasParams = exports.getEthTxCallParams = exports.getSendTxParams = void 0;
const web3_errors_1 = require("web3-errors");
const web3_utils_1 = require("web3-utils");
const encoding_1 = require("./encoding");
const getSendTxParams = ({ abi, params, options, contractOptions, }) => {
    var _a;
    const deploymentCall = (_a = options === null || options === void 0 ? void 0 : options.data) !== null && _a !== void 0 ? _a : contractOptions.data;
    if (!deploymentCall && !(options === null || options === void 0 ? void 0 : options.to) && !contractOptions.address) {
        throw new web3_errors_1.Web3ContractError('Contract address not specified');
    }
    if (!(options === null || options === void 0 ? void 0 : options.from) && !contractOptions.from) {
        throw new web3_errors_1.Web3ContractError('Contract "from" address not specified');
    }
    let txParams = (0, web3_utils_1.mergeDeep)({
        to: contractOptions.address,
        gas: contractOptions.gas,
        gasPrice: contractOptions.gasPrice,
        from: contractOptions.from,
        data: contractOptions.data,
    }, options);
    if (!txParams.data || abi.type === 'constructor') {
        txParams = Object.assign(Object.assign({}, txParams), { data: (0, encoding_1.encodeMethodABI)(abi, params, txParams.data) });
    }
    return txParams;
};
exports.getSendTxParams = getSendTxParams;
const getEthTxCallParams = ({ abi, params, options, contractOptions, }) => {
    if (!(options === null || options === void 0 ? void 0 : options.to) && !contractOptions.address) {
        throw new web3_errors_1.Web3ContractError('Contract address not specified');
    }
    let txParams = (0, web3_utils_1.mergeDeep)({
        to: contractOptions.address,
        gas: contractOptions.gas,
        gasPrice: contractOptions.gasPrice,
        from: contractOptions.from,
        data: contractOptions.data,
    }, options);
    txParams = Object.assign(Object.assign({}, txParams), { data: (0, encoding_1.encodeMethodABI)(abi, params, txParams.data ? (0, web3_utils_1.toHex)(txParams.data) : undefined) });
    return txParams;
};
exports.getEthTxCallParams = getEthTxCallParams;
const getEstimateGasParams = ({ abi, params, options, contractOptions, }) => {
    let txParams = (0, web3_utils_1.mergeDeep)({
        to: contractOptions.address,
        gas: contractOptions.gas,
        gasPrice: contractOptions.gasPrice,
        from: contractOptions.from,
        data: contractOptions.data,
    }, options);
    txParams = Object.assign(Object.assign({}, txParams), { data: (0, encoding_1.encodeMethodABI)(abi, params, txParams.data ? (0, web3_utils_1.toHex)(txParams.data) : undefined) });
    return txParams;
};
exports.getEstimateGasParams = getEstimateGasParams;
const isContractInitOptions = (options) => typeof options === 'object' &&
    !(0, web3_utils_1.isNullish)(options) &&
    [
        'data',
        'from',
        'gas',
        'gasPrice',
        'gasLimit',
        'address',
        'jsonInterface',
        'syncWithContext',
    ].some(key => key in options);
exports.isContractInitOptions = isContractInitOptions;
const isWeb3ContractContext = (options) => typeof options === 'object' && !(0, web3_utils_1.isNullish)(options) && !(0, exports.isContractInitOptions)(options);
exports.isWeb3ContractContext = isWeb3ContractContext;
const getCreateAccessListParams = ({ abi, params, options, contractOptions, }) => {
    if (!(options === null || options === void 0 ? void 0 : options.to) && !contractOptions.address) {
        throw new web3_errors_1.Web3ContractError('Contract address not specified');
    }
    if (!(options === null || options === void 0 ? void 0 : options.from) && !contractOptions.from) {
        throw new web3_errors_1.Web3ContractError('Contract "from" address not specified');
    }
    let txParams = (0, web3_utils_1.mergeDeep)({
        to: contractOptions.address,
        gas: contractOptions.gas,
        gasPrice: contractOptions.gasPrice,
        from: contractOptions.from,
        data: contractOptions.data,
    }, options);
    if (!txParams.data || abi.type === 'constructor') {
        txParams = Object.assign(Object.assign({}, txParams), { data: (0, encoding_1.encodeMethodABI)(abi, params, txParams.data) });
    }
    return txParams;
};
exports.getCreateAccessListParams = getCreateAccessListParams;
